packages/rlc-common/src/buildSamples.ts

```diff
@@ -0,0 +1,835 @@
+  path: Set<string> = new Set()
+) {
+  const schema = schemaMap.get(type);
+  if (schema && schema.enum && schema.enum.length > 0) {
+    const first = schema.enum[0];
+    const itemType = schema.typeName ?? schema.type;
+    if (itemType === "string" || (first.typeName ?? first.type) === undefined) {
+      return `"${first}"`;
+    } else if (itemType === "number") {
+      return `${first}`;
+    }
+    return mockParameterTypeValue(
+      first.typeName ?? first.type ?? first,
+      parameterName,
+      schemaMap,
+      path
+    );
+  }
+  const unionType = getUnionType(type);
+  return mockParameterTypeValue(unionType, parameterName, schemaMap, path);
+}
+
+function mockRecordValues(
+  type: string,
+  parameterName: string,
+  schemaMap: Map<string, Schema>,
+  path: Set<string> = new Set()
+) {
+  let recordType;
+  const schema = schemaMap.get(type) as DictionarySchema;
+  if (schema && schema.additionalProperties) {
+    recordType =
+      schema.additionalProperties.typeName ?? schema.additionalProperties.type;
+    if (recordType !== "string" && !schemaMap.has(recordType)) {
+      schemaMap.set(recordType, schema.additionalProperties);
+    }
+  } else {
+    recordType = getRecordType(type);
+  }
+
+  return recordType
+    ? `{"key": ${mockParameterTypeValue(
+        recordType,
+        parameterName,
+        schemaMap,
+        path
+      )}}`
+    : `{}`;
+}
+
+function mockArrayValues(
+  type: string,
+  parameterName: string,
+  schemaMap: Map<string, Schema>,
+  path: Set<string> = new Set()
+) {
+  let arrayType;
+  const schema = schemaMap.get(type) as ArraySchema;
+  if (schema && schema.items) {
+    arrayType = schema.items.typeName ?? schema.items.type;
+    if (arrayType !== "string" && !schemaMap.has(arrayType)) {
+      schemaMap.set(arrayType, schema.items);
+    }
+  } else if (isArrayObject(type)) {
+    arrayType = getArrayObjectType(type);
+  } else if (isNativeArray(type)) {
+    arrayType = getNativeArrayType(type);
+  }
+  const itemValue = arrayType
+    ? mockParameterTypeValue(arrayType, parameterName, schemaMap, path)
+    : undefined;
+  return itemValue ? `[${itemValue}]` : `[]`;
+}
+
+function mockObjectValues(
+  type: string,
+  _parameterName: string,
+  schemaMap: Map<string, Schema>,
+  path: Set<string> = new Set()
+) {
+  if (path.has(type)) {
+    // skip generating if self referenced
+    return `{} as any /**FIXME */`;
+  }
+  path.add(type);
+  // add object's properties
+  const values: string[] = [];
+  const schema = schemaMap.get(type) as ObjectSchema;
+  const allPropNames = new Set<string>();
+  extractObjectProperties(
+    schema?.properties ?? {},
+    schemaMap,
+    allPropNames,
+    values,
+    path
+  );
+
+  // add parents' properties
+  const parents = getAllParents(schema, schemaMap);
+  for (let parent of parents) {
+    extractObjectProperties(
+      (parent as ObjectSchema)?.properties ?? {},
+      schemaMap,
+      allPropNames,
+      values,
+      path
```

Comment:
Can we get all properties first and then extractObjectProperties ? That way we don't need the allPropNames to decide if a property has been process before and we don't have to call this function multiple times ?

Reply:
Sure